src/enzyme_ad/jax/Passes/OptimizeCommunication.cpp

```diff
@@ -1337,6 +1335,106 @@ struct WrapToPadCommOptimize : public OpRewritePattern<enzymexla::WrapOp> {
   }
 };
 
+// Communication optimization pattern to rewrite wrap operations in terms of
+// pad and rotate operations. This can enable better optimization of the
+// resulting operations in distributed/sharded contexts.
+//
+// Pattern: wrap(x, lhs=L, rhs=R) =>
+//   p0 = pad(x, lhs=L, rhs=R)
+//   p1 = rotate(p0, L)
+//   p2 = rotate(p0, -R)
+//   result = select based on iota to choose between p1 (left), p0 (middle), p2
+//   (right)
+//
+// Note: Only applied when no sharding is present, to be conservative.
+// This pattern is intended to work in conjunction with other communication
+// optimization patterns for pad and rotate operations.
+struct WrapToRotateOptimize : public OpRewritePattern<enzymexla::WrapOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(enzymexla::WrapOp wrap,
+                                PatternRewriter &rewriter) const override {
+    // Check if already inside a manual computation
+    if (wrap->getParentOfType<sdy::ManualComputationOp>())
+      return failure();
+
+    // Only apply if there's no sharding (for now, to be conservative)
+    if (mlir::sdy::getSharding(wrap))
+      return failure();
```

Comment:
I think this is wrong, we instead only want to apply if there are multiple devices along the relevant axis, x/ref: https://github.com/EnzymeAD/Enzyme-JAX/blob/04c15dc7c3736cca6f0a93c9934829fde489cc11/src/enzyme_ad/jax/Passes/OptimizeCommunication.cpp#L1377